### PR TITLE
When sourcekitd crashes, log the file contents with which it crashed and the request

### PR DESCRIPTION
**Pull request by ahoppen**
This should make it a lot easier to reproduce sourcekitd crashes.

### CODE REVIEW

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 1]
```diff
@@ -620,12 +620,7 @@ extension SourceKitServer: MessageHandler {
   private func handleImpl(_ params: some NotificationType, from clientID: ObjectIdentifier) async {
     let notification = Notification(params, clientID: clientID)
 
-    logger.log(
-      """
-      Received notification
-      \(notification.forLogging)
-      """
-    )
+    logger.log("Received notification: \(notification.forLogging)")
```
Comment (bnbarham) [inline Sources/SourceKitLSP/SourceKitServer.swift:623]: Presumably you found this looks nicer without the newline?

File: Sources/SourceKitD/SourceKitD.swift [hunk 2]
```diff
@@ -107,9 +129,7 @@ extension SourceKitD {
 }
 
 private func logRequest(_ request: SKDRequestDictionary) {
-  // FIXME: Ideally we could log the request key here at the info level but the dictionary is
-  // readonly.
-  logger.log(
+  logger.info(
```
Comment (bnbarham) [inline Sources/SourceKitD/SourceKitD.swift:132]: What's default vs info? The docs weren't particularly illuminating.

**Review by bnbarham — COMMENTED** [review]
Should we log similar for clangd?

File: Sources/SourceKitD/SourceKitD.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SourceKitD/SourceKitD.swift:132]: https://github.com/apple/sourcekit-lsp/blob/78c28f5dec294ffa87ffcbd91fb9bbeae9d8cbd4/Sources/LSPLogging/Logging.swift#L13-L18

File: Sources/SourceKitLSP/SourceKitServer.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitServer.swift:623]: Yes, my use case for this was that I could grep for `Received notification` and see the types of all the notifications that got sent.

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SourceKitD/SourceKitD.swift [hunk 3]
```diff
@@ -96,6 +100,24 @@ extension SourceKitD {
...
+          """
+        let chunks = splitLongMultilineMessage(message: log)
+        for (index, chunk) in chunks.enumerated() {
+          logger.fault(
+            """
+            sourcekitd crashed (\(index + 1)/\(chunks.count))
+            \(chunk)
+            """
+          )
+        }
```
Comment (hamishknight) [inline Sources/SourceKitD/SourceKitD.swift:119]: If OSLog treats this as private (which I believe is the default?), will we just log multiple `sourcekitd crashed` lines with redacted chunks?

File: Sources/SourceKitD/SourceKitD.swift (hunk 2 above)
Comment (bnbarham) [reply Sources/SourceKitD/SourceKitD.swift:129]: Ah, so persisted vs not 👍

File: Sources/SourceKitD/SourceKitD.swift (hunk 3 above)
Comment (ahoppen) [reply Sources/SourceKitD/SourceKitD.swift:119]: Yes, we will. It’s a big unfortunate but I don’t have any better idea.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**